File: Problem2.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
public class Codec {
    int idx;

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root == null)
            return "";
        StringBuilder sb = new StringBuilder();
        sb.append(root.val);
        serializeHelper(root,sb);
        System.out.println(sb);
        return sb.toString();
    }
    private void serializeHelper(TreeNode root, StringBuilder sb){
        //base

        //logic
        if(root.left!=null){
            sb.append(",");
            sb.append(root.left.val);
            serializeHelper(root.left,sb);
        }
        else{
            sb.append(",");
            sb.append("#");
        }
        if(root.right!=null){
            sb.append(",");
            sb.append(root.right.val);
            serializeHelper(root.right,sb);
        }
        else{
            sb.append(",");
            sb.append("#");
        }
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data.equals(""))
            return null;
        String[] strArr = data.split(",");
        TreeNode root = new TreeNode(Integer.parseInt(strArr[0]));
        this.idx = 1;
        deserializeHelper(root,strArr);
        return root;
    }
    private void deserializeHelper(TreeNode root, String[] strArr){
        if(!strArr[idx].equals("#")){
            root.left = new TreeNode(Integer.parseInt(strArr[idx]));
            idx++;
            deserializeHelper(root.left,strArr);
        }
        else{
            idx++;
        }
        if(!strArr[idx].equals("#")){
            root.right = new TreeNode(Integer.parseInt(strArr[idx]));
            idx++;
            deserializeHelper(root.right,strArr);
        }
        else{
            idx++;
        }
    }
}

File: Problem2.py
#Time Complexity :O(n)
#Space Complexity :O(n)
#Did this code successfully run on Leetcode : Yes
#Any problem you faced while coding this : No

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
#DFS solution is complicated as it involves merge sort, we need to keep track of the depth also so BFS is better soltuion
from collections import deque
class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        q=deque()
        colq=deque()
        q.append(root)
        colq.append(0)
        map={}
        minVar=0
        maxVar=0
        result=[]
        while q:
            curr=q.popleft()
            currCol=colq.popleft()
            if currCol not in map.keys():
                map[currCol]=[]
            map.get(currCol).append(curr.val)
            if curr.left:
                q.append(curr.left)
                colq.append(currCol-1)
                minVar=min(minVar,currCol-1)
            if curr.right:
                q.append(curr.right)
                colq.append(currCol+1)
                maxVar=max(maxVar,currCol+1)

        for i in range(minVar,maxVar+1):
            result.append(map.get(i))

        return result